lib/ansible/modules/storage/glusterfs/gluster_peer.py

```diff
@@ -0,0 +1,169 @@
+
+RETURN = '''
+'''
+
+from ansible.module_utils.basic import AnsibleModule
+from distutils.version import LooseVersion
+
+
+class Peer(object):
+    def __init__(self, module):
+        self.module = module
+        self.state = self.module.params['state']
+
+    def gluster_peer_ops(self):
+        nodes = self.module.params['nodes']
+        if not nodes:
+            self.module.fail_json(msg="nodes list cannot be empty")
+        force = 'force' if self.module.params.get('force') else ''
+        if self.state == 'present':
+            nodes = self.get_to_be_probed_hosts(nodes)
+            action = 'probe'
+            # In case of peer probe, we do not need `force'
+            force = ''
+        else:
+            action = 'detach'
+        cmd = []
+        for node in nodes:
+            cmd.append(' peer ' + action + ' ' + node + ' ' + force)
+        return cmd
+
+    def get_to_be_probed_hosts(self, hosts):
+        rc, output, err = self.module.run_command("gluster pool list")
+        peers_in_cluster = [line.split('\t')[1].strip() for
+                            line in filter(None, output.split('\n')[1:])]
+        try:
+            peers_in_cluster.remove('localhost')
+        except ValueError:
+            # It is ok not to have localhost in list
+            pass
+        hosts_to_be_probed = [host for host in hosts if host not in
+                              peers_in_cluster]
+        return hosts_to_be_probed
+
+    def call_peer_commands(self, cmds):
+        errors = []
+        for cmd in cmds:
+            rc, out, err = self._run_command('gluster', cmd)
+            if rc:
+                errors.append(err)
+        return errors
+
+    def get_output(self, errors):
+        if not errors:
+            self.module.exit_json(rc=0, changed=True)
+        else:
+            self.module.fail_json(rc=1, msg='\n'.join(errors))
+
+    def _run_command(self, op, opts):
+        cmd = self.module.get_bin_path(op, True) + opts + ' --mode=script'
+        return self.module.run_command(cmd)
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            force=dict(type='bool', required=False),
+            nodes=dict(type='list', required=True),
+            state=dict(type='str', choices=['absent', 'present'],
+                       default='present'),
+        ),
+        supports_check_mode=False
+    )
+    pops = Peer(module)
+    required_version = "3.2"
+    # Verify if required GlusterFS version is installed
+    if is_invalid_gluster_version(module, required_version):
+        module.fail_json(msg="GlusterFS version > %s is required" %
+                         required_version)
+    cmds = pops.gluster_peer_ops()
+    errors = pops.call_peer_commands(cmds)
+    pops.get_output(errors)
+
+
+def is_invalid_gluster_version(module, required_version):
+    cmd = module.get_bin_path('gluster', True) + ' --version'
+    result = module.run_command(cmd)
+    ver_line = result[1].split('\n')[0]
```

Comment:
Will this parsing also work when gluster isn't installed at all? In that case any meaningfull message would be nice.

Reply:
Also you may get here(or on many other places) an exception, and you don't catch any. AFAIK Ansible modules should return always with `module.exit_json` or `module.fail_json` rather then raising `Exception`.

Reply:
http://docs.ansible.com/ansible/latest/dev_guide/developing_modules_general.html#new-module-development

There is written:
```
# during the execution of the module, if there is an exception or a
    # conditional state that effectively causes a failure, run
    # AnsibleModule.fail_json() to pass in the message and the result
    if module.params['name'] == 'fail me':
        module.fail_json(msg='You requested this to fail', **result)
```

AFAIK module should return with `module.exit_json` or `module.fail_json`, not `raise ValueError` for example.

Reply:
I meant: where `ValueError` could be raised ?

Reply:
I meant `IndexError` not `ValueError`, but it don't really matter. What I meant is, that isn't it better to have this safe contruct there:

```python
try:
  main()
except Exception as e:
  module.fail_json(msg=str(e), exception=traceback.format_exc())
```